File: javascript/bracket-push/bracket-push.js
const pairs = {
  '{' : '}',
  '[' : ']',
  '(' : ')'
};

module.exports = function(brackets) {
  let queue = [];
  [...brackets].forEach((bracket) => {
    if (pairs.hasOwnProperty(bracket)) {
      queue.push(bracket);
    } else if (pairs[queue[queue.length-1]] == bracket) {
      queue.pop();
    } else {
      queue.push(false);
    }
  });
  return queue.length == 0;
};

File: ruby/bracket-push/bracket_push.rb
module Brackets
  @@pairs = {
    '{' => '}',
    '[' => ']',
    '(' => ')'
  }

  @@brackets_only = /[^\[\{\(\]\}\)]/

  def self.paired?(input)
    queue = []
    input.gsub(@@brackets_only, '').chars do |b|
      if @@pairs.keys.include?(b)
        queue.push(b)
      elsif @@pairs[queue.last] == b
        queue.pop
      else
        queue.push(nil)
      end
    end
    queue.length == 0
  end
end

module BookKeeping
  VERSION = 2
end
